### PR TITLE
Chore: switch src/documents/bulk*.py from os.path to pathlib.Path

### PR DESCRIPTION
**Pull request by gothicVI**
## Proposed change

Switch `src/documents/bulk*.py` from `os.path` to `pathlib.Path`.

This commit strictly preserves the return types.

Related to https://github.com/paperless-ngx/paperless-ngx/discussions/7861

## Type of change

- [ ] Bug fix: non-breaking change which fixes an issue.
- [ ] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [x] Other. Please explain: Switch `src/documents/bulk*.py` from `os.path` to `pathlib.Path`.

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [ ] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [ ] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by stumpylog** [comment]
Yes, breaking it into smaller chunks is preferred

**Comment by shamoon** [comment]
Yea, and without sounding rude, it strikes me as a bit of an odd contribution, like perhaps something better left to core maintainers of a project?

**Comment by gothicVI** [comment]
@shamoon not sounding rude at all. I just saw it, like the project, and wanted to contribute.
I know python but not TypeScript so that would be my only option ;)

If I may ask though: why would it be? Docs and linting/formatting or other chores are quite usual first contact points for people to add to FOSS projects.

**Comment by stumpylog** [comment]
I'm fine with you working on it, but a change set of now 45 files isn't really reviewable

**Comment by shamoon** [comment]
> Docs and linting/formatting or other chores are quite usual first contact points for people to add to FOSS projects.

I suppose though not usually in the form of such a large diff.

Anyway, I’ll stay out of it mostly.

**Comment by gothicVI** [comment]
@stumpylog I've reduced the PR to a small subset. Is that OK?

**Comment by stumpylog** [comment]
I won't be near a computer to review today, but the count looks better, thanks

**Review by stumpylog — CHANGES_REQUESTED** [review]
I'm not seeing the point to these changes.  You convert something to a Path, then immediately convert back to a string.  That defeats the purpose of using a Path object.

It would be better to take a few functions at a time, trace usages and make sure everything supports a Path as needed.

**Comment by stumpylog** [comment]
I'll have a look after this beta.  It's seemed stable and no reason to upset that

File: .ruff.toml [hunk 1]
```diff
@@ -31,17 +31,61 @@ extend-select = [
   "PLE",   # https://docs.astral.sh/ruff/rules/#pylint-pl
   "RUF",   # https://docs.astral.sh/ruff/rules/#ruff-specific-rules-ruf
   "FLY",   # https://docs.astral.sh/ruff/rules/#flynt-fly
+  "PTH",   # https://docs.astral.sh/ruff/rules/#flake8-use-pathlib-pth
 ]
-# TODO PTH https://docs.astral.sh/ruff/rules/#flake8-use-pathlib-pth
 ignore = ["DJ001", "SIM105", "RUF012"]
 
 [lint.per-file-ignores]
```
[the remark this answers is not in the document]
Comment (gothicVI) [reply .ruff.toml:38]: Probably yes but then any new file wouldn't be covered from the get go

File: src/documents/bulk_download.py [hunk 2]
```diff
@@ -37,32 +44,38 @@ def _formatted_filepath(
...
+    ) -> Path:
         """
         Constructs a full file path for the given document to be used inside
         the zipfile.
 
         The path is already unique, as handled when a document is consumed or updated
         """
         if archive and doc.has_archive_version:
-            in_archive_path = os.path.join(folder, doc.archive_filename)
+            if doc.archive_filename is None:
```
Comment (stumpylog) [inline src/documents/bulk_download.py:55]: This line cannot be hit, `has_archive_version` checks the condition of `archive_filename is not None` already.

If this is just for typing, you could use `TYPE_CHECKING` and assert the filename is not None there

File: src/documents/bulk_download.py [hunk 3]
```diff
@@ -37,32 +44,38 @@ def _formatted_filepath(
...
+                raise ValueError("doc %a has no archive_filename", doc)
+            in_archive_path: Path = Path(folder) / doc.archive_filename
         else:
-            in_archive_path = os.path.join(folder, doc.filename)
+            if doc.filename is None:
```
Comment (stumpylog) [inline src/documents/bulk_download.py:59]: I don't think the `filename` of a model can ever be None either, even though the DB column is nullable, it is always assigned to something or we'd lose the file.

File: src/documents/bulk_download.py (hunk 2 above)
Comment (gothicVI) [reply src/documents/bulk_download.py]: That's a perfect solution!

File: src/documents/bulk_edit.py [hunk 4]
```diff
@@ -321,31 +344,35 @@ def split(
...
+    user: User | None = None,
+) -> Literal["OK"]:
     logger.info(
         f"Attempting to split document {doc_ids[0]} into {len(pages)} documents",
     )
     doc = Document.objects.get(id=doc_ids[0])
     import pikepdf
 
-    consume_tasks = []
+    consume_tasks: list = []
```
Comment (stumpylog) [inline src/documents/bulk_edit.py:355]: Again with celery, this probably doesn't help type anything since the signature is not really used directly anyway

**Review by stumpylog — CHANGES_REQUESTED** [review]
Overall looks good.  i think there's some redundant typing added, and some that don't really add value.  See what you think

**Review by stumpylog — APPROVED** [review]

**Pull request merged**